File: csphn_hsg2023_maxsqr.py
m, n, k = tuple(int(i) for i in input().split())

matrix = [[0 for _ in range(n)] for _ in range(m)]

def calcsquare(matrix):
    dp = [[0 for _ in range(n + 5)] for _ in range(m + 5)]

    for i in range(m):
        dp[i][0] = matrix[i][0]
    for i in range(n):
        dp[0][i] = matrix[0][i]

    for i in range(1, m):
        for j in range(1, n):
            if matrix[i][j] == matrix[i][j - 1] == matrix[i - 1][j] == matrix[i - 1][j - 1]:
                dp[i][j] = min(dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]) + 1
            else:
                dp[i][j] = 0

    maxdp = max(max(row) for row in dp)
    return maxdp

for _ in range(k):
    t, l, b, r = tuple(int(i) - 1 for i in input().split())
    for i in range(t, b + 1):
        for j in range(l, r + 1):
            matrix[i][j] += 1
for i in matrix:
    print(*i)
print(calcsquare(matrix))

File: csphn_hsg2023_maxsqr.cpp
#include<bits/stdc++.h>

using namespace std;
const int maxn = (int) 2e3 + 5;
int a[maxn][maxn], dp[maxn][maxn];
signed main() {
    int m, n, k;
    cin >> m >> n >> k;
    for(int i = 1; i <= m; i++) {
        for(int j = 1; j <= n; j++) {
            a[i][j] = 0;
            dp[i][j] = 0;
        }
    }
    for(int _ = 0; _ < k; _++) {
        int t, l, b, r;
        cin >> t >> l >> b >> r;
        a[t][l]++;
        a[b + 1][r + 1]++;
        a[b + 1][l]--;
        a[t][r + 1]--;
    }
    for(int i = 1; i <= m; i++) {
        for(int j = 1; j <= n; j++) {
            a[i][j] += a[i - 1][j];
        }
    }
    for(int i = 1; i <= m; i++) {
        for(int j = 1; j <= n; j++) {
            a[i][j] += a[i][j - 1];
        }
    }
    for(int i = 1; i <= m; i++) {
        for(int j = 1; j <= n; j++) {
            if(a[i][j] == a[i - 1][j] and a[i][j] == a[i][j - 1] and a[i][j] == a[i - 1][j - 1]) {
                dp[i][j] = min(dp[i - 1][j], min(dp[i][j - 1], dp[i - 1][j - 1])) + 1;
            } else dp[i][j] = 1;
        }
    }/*
    for(int i = 1; i <= n; i++) {
        for(int j = 1; j <= n; j++) {
            cout << dp[i][j] << " ";
        }
        cout << endl;
    }*/
    int maxdp = 0;
    for(int i = 1; i <= m; i++) {
        for(int j = 1; j <= n; j++) {
            maxdp = max(maxdp, dp[i][j]);
        }
    }
    cout << maxdp;
}